src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,407 @@
+			logger.info("Lost lease, so terminating.");
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void shardEnded(ShardEndedInput shardEndedInput) {
+			try {
+				logger.info("Reached shard end checkpointing.");
+				shardEndedInput.checkpointer().checkpoint();
+			}
+			catch (ShutdownException | InvalidStateException e) {
+				logger.error("Exception while checkpointing at shard end.  Giving up", e);
+			}
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void shutdownRequested(ShutdownRequestedInput shutdownRequestedInput) {
+			try {
+				logger.info("Scheduler is shutting down, checkpointing.");
+				shutdownRequestedInput.checkpointer().checkpoint();
+			}
+			catch (ShutdownException | InvalidStateException e) {
+				logger.error("Exception while checkpointing at requested shutdown.  Giving up", e);
+			}
+		}
+
+		/**
+		 * Process records performing retries as needed. Skip "poison pill" records.
+		 *
+		 * @param records Data records to be processed.
+		 */
+		private void processRecordsWithRetries(List<KinesisClientRecord> records) {
+			for (KinesisClientRecord record : records) {
+				try {
+					processSingleRecord(record);
+				}
+				catch (Throwable t) {
+					logger.warn("Caught throwable while processing record " + record, t);
+				}
+				finally {
+					attributesHolder.remove();
+				}
+			}
+		}
+
+		/**
+		 * Process a single record.
+		 *
+		 * @param record The record to be processed.
+		 */
+		private void processSingleRecord(KinesisClientRecord record) {
+
+			// Convert AWS Record in Spring Message.
+			performSend(prepareMessageForRecord(record), record);
+		}
+
+		private void performSend(AbstractIntegrationMessageBuilder<?> messageBuilder, Object rawRecord) {
+			Message<?> messageToSend = messageBuilder.build();
+			setAttributesIfNecessary(rawRecord, messageToSend);
+			try {
+				sendMessage(messageToSend);
+			}
+			catch (Exception e) {
+				logger.error("Got an exception during sending a '" + messageToSend + "'" + "\nfor the '" + rawRecord
+						+ "'.\n" + "Consider to use 'errorChannel' flow for the compensation logic.", e);
+			}
+		}
+
+		/**
+		 * If there's an error channel, we create a new attributes holder here.
+		 * Then set the attributes for use by the {@link ErrorMessageStrategy}.
+		 * @param record the Kinesis record to use.
+		 * @param message the Spring Messaging message to use.
+		 */
+		private void setAttributesIfNecessary(Object record, Message<?> message) {
+			if (getErrorChannel() != null) {
+				AttributeAccessor attributes = ErrorMessageUtils.getAttributeAccessor(message, null);
+				attributesHolder.set(attributes);
+				attributes.setAttribute(AwsHeaders.RAW_RECORD, record);
+			}
+		}
+
+		private AbstractIntegrationMessageBuilder<Object> prepareMessageForRecord(KinesisClientRecord record) {
+			ByteBuffer data = record.data();
+			byte[] dataArray = new byte[data.remaining()];
+			Object payload = dataArray;
+			data.get(dataArray);
```

Comment:
Why just `data.array()` is not enough?

Reply:
I just tried, and got some ReadOnlyBufferException so I think we'll have to keep using a copy.